File: Web/Helpers/CuilHelper.cs
namespace API.Helpers;

/// <summary>
/// Helpers para trabajar con CUILs
/// </summary>
public static class CuilHelper
{
    /// <summary>
    /// Normaliza un CUIL removiendo guiones y espacios
    /// </summary>
    /// <param name="cuil">CUIL con o sin guiones</param>
    /// <returns>CUIL solo con dígitos</returns>
    public static string Normalize(string cuil)
    {
        if (string.IsNullOrWhiteSpace(cuil))
            return cuil;

        return cuil.Replace("-", "").Replace(" ", "").Trim();
    }

    /// <summary>
    /// Formatea un CUIL agregando guiones
    /// </summary>
    /// <param name="cuil">CUIL sin guiones</param>
    /// <returns>CUIL formateado XX-XXXXXXXX-X</returns>
    public static string Format(string cuil)
    {
        if (string.IsNullOrWhiteSpace(cuil))
            return cuil;

        var normalized = Normalize(cuil);

        if (normalized.Length != 11)
            return cuil; // Retornar original si no tiene 11 dígitos

        return $"{normalized.Substring(0, 2)}-{normalized.Substring(2, 8)}-{normalized.Substring(10, 1)}";
    }
}

File: cliente/src/utils/cuilHelper.js
/**
 * Helper para trabajar con CUILs (Código Único de Identificación Laboral)
 */

/**
 * Normaliza un CUIL removiendo guiones y espacios
 * @param {string} cuil - CUIL con o sin formato
 * @returns {string} CUIL solo con dígitos
 */
export const normalizeCuil = (cuil) => {
    if (!cuil) return '';
    return cuil.replace(/-/g, '').replace(/\s/g, '').trim();
};

/**
 * Formatea un CUIL agregando guiones en formato XX-XXXXXXXX-X
 * @param {string} cuil - CUIL sin formato o con formato
 * @returns {string} CUIL formateado
 */
export const formatCuil = (cuil) => {
    if (!cuil) return '';
    
    const normalized = normalizeCuil(cuil);
    
    // Si ya tiene el formato correcto con guiones, devolverlo
    if (cuil.includes('-') && cuil.length === 13) return cuil;
    
    // Si no tiene 11 dígitos, devolver el original
    if (normalized.length !== 11) return cuil;
    
    // Formatear: XX-XXXXXXXX-X
    return `${normalized.substring(0, 2)}-${normalized.substring(2, 10)}-${normalized.substring(10)}`;
};

/**
 * Valida el formato básico de un CUIL (sin validar dígito verificador)
 * Esta validación es para el frontend. El backend hace la validación completa.
 * @param {string} cuil - CUIL a validar
 * @returns {boolean} true si el formato es válido
 */
export const validateCuilFormat = (cuil) => {
    if (!cuil) return false;
    
    const normalized = normalizeCuil(cuil);
    
    // Debe tener exactamente 11 dígitos
    if (normalized.length !== 11) return false;
    
    // Debe contener solo números
    return /^\d{11}$/.test(normalized);
};

/**
 * Valida un CUIL completo incluyendo el dígito verificador
 * @param {string} cuil - CUIL a validar
 * @returns {boolean} true si el CUIL es válido
 */
export const validateCuil = (cuil) => {
    if (!validateCuilFormat(cuil)) return false;
    
    const normalized = normalizeCuil(cuil);
    const multiplicadores = [5, 4, 3, 2, 7, 6, 5, 4, 3, 2];
    
    let suma = 0;
    for (let i = 0; i < 10; i++) {
        suma += parseInt(normalized[i]) * multiplicadores[i];
    }
    
    const resto = suma % 11;
    const digitoVerificador = resto === 0 ? 0 : resto === 1 ? 9 : 11 - resto;
    
    return digitoVerificador === parseInt(normalized[10]);
};
